engine/utils/collections/StackArray.h

```diff
@@ -0,0 +1,480 @@
+        ArrayUtils::AssertIsInBounds(index, S);
+        return data[index];
+    }
+
+    /**
+     * Gets a mutable reference to a value stored in the specified index.
+     * @param index the index position to search.
+     * @return a constant reference to the value stored in the index.
+     */
+    inline T& Get(const size_t& index)
+    {
+        return data[index];
+    }
+
+    /**
+     * @brief Returns the const raw pointer stored by the array.
+     * @return the raw pointer stored by the array.
+     */
+    inline const T* Data() const
+    {
+        return data;
+    }
+
+    /**
+     * @brief Returns the raw pointer stored by the array.
+     * @return the raw pointer stored by the array.
+     */
+    inline T* Data()
+    {
+        return data;
+    }
+
+    /**
+     * @brief Returns the size of the array in bytes.
+     * @return the size of the array in bytes.
+     */
+    inline constexpr const size_t Size() const
+    {
+        return S;
+    }
+
+    /**
+     * @brief Returns the size of the array in bytes.
+     * @return the size of the array in bytes.
+     */
+    inline const size_t Count() const
+    {
+        return count;
+    }
+
+    // Setters
+
+    /**
+     * @brief Inserts a given element into the position specified by `index`. If the provided index
+     * is larger than the array's size, an exception will be thrown.
+     * @param index the position in the array we want to insert the element into.
+     * @param element the element we want to store in the array.
+     */
+    inline void Insert(const size_t& index, const T& element)
+    {
+        count += !bitField.IsSet(index + 1);
+        bitField.SetBit(index + 1);
+        Set(index, element);
+    }
+
+    inline void Append(const T& element)
+    {
+        ArrayUtils::AssertIsInBounds(count, S);
+        Set(count, element);
+        count += !bitField.IsSet(count + 1);
+        bitField.SetBit(count);
+    }
+
+    /**
+     * Checks if the MSArray is empty.
+     * @return true if the MSArray has had no allocations made, false if at least one element was
+     * added.
+     */
+    inline bool Empty() const
+    {
+        return count == 0;
+    }
+
+    /**
+     * Checks if a specified index has been assigned to. If the provided index
+     * is larger than the array's size, an exception will be thrown.
+     * @param index the index of the element being searched for.
+     * @return true if the index has been previously assigned to.
+     */
+    inline bool Active(const size_t& index) const
+    {
+        return bitField.IsSet(index + 1);
+    }
+
+    /**
+     * @brief Removes the element at the specified position from the array. If the provided index
+     * is larger than the array's size, an exception will be thrown.
+     * @param index the index of the element we want to remove.
+     */
+    inline void Remove(const size_t& index)
+    {
+        bitField.UnsetBit(index + 1);
+        count--;
```

Comment:
For safety reasons always prefix increment/decrement operators unless you explicitly want that functionality. There's been a few such bugs related to this so far.

Reply:
Changed.